// @fluencelabs/avm/dist/avmHelpers.js
"use strict";
/*
 * Copyright 2022 Fluence Labs Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.callAvm = exports.deserializeAvmResult = exports.serializeAvmArgs = void 0;
var decoder = new TextDecoder();
var encoder = new TextEncoder();
/**
 * Encodes arguments into JSON array suitable for marine-js
 * @param initPeerId - peer ID which initialized particle
 * @param currentPeerId - peer ID which is currently executing the particle
 * @param air - particle's air script as string
 * @param prevData - particle's prev data as raw byte array
 * @param data - particle's data as raw byte array
 * @param callResults - array of tuples [callResultKey, callResult]
 * @returns AVM call arguments suitable for marine-js
 */
function serializeAvmArgs(runParams, air, prevData, data, callResults) {
    var callResultsToPass = {};
    for (var _i = 0, callResults_1 = callResults; _i < callResults_1.length; _i++) {
        var _a = callResults_1[_i], key = _a[0], callResult = _a[1];
        callResultsToPass[key] = {
            ret_code: callResult.retCode,
            result: callResult.result,
        };
    }
    var encoded = encoder.encode(JSON.stringify(callResultsToPass));
    return [
        // force new line
        air,
        Array.from(prevData),
        Array.from(data),
        {
            init_peer_id: runParams.initPeerId,
            current_peer_id: runParams.currentPeerId,
            timestamp: runParams.timestamp,
            ttl: runParams.ttl,
        },
        Array.from(encoded),
    ];
}
exports.serializeAvmArgs = serializeAvmArgs;
/**
 * Deserializes raw result of AVM call obtained from marine-js into structured form
 * @param rawResult - string containing raw result of AVM call
 * @returns structured InterpreterResult
 */
function deserializeAvmResult(result) {
    var callRequestsStr = decoder.decode(new Uint8Array(result.call_requests));
    var parsedCallRequests;
    try {
        if (callRequestsStr.length === 0) {
            parsedCallRequests = {};
        }
        else {
            parsedCallRequests = JSON.parse(callRequestsStr);
        }
    }
    catch (e) {
        throw "Couldn't parse call requests: " + e + '. Original string is: ' + callRequestsStr;
    }
    var resultCallRequests = [];
    for (var key in parsedCallRequests) {
        var callRequest = parsedCallRequests[key];
        var arguments_ = void 0;
        var tetraplets = void 0;
        try {
            arguments_ = JSON.parse(callRequest.arguments);
        }
        catch (e) {
            throw "Couldn't parse arguments: " + e + '. Original string is: ' + arguments_;
        }
        try {
            tetraplets = JSON.parse(callRequest.tetraplets);
        }
        catch (e) {
            throw "Couldn't parse tetraplets: " + e + '. Original string is: ' + tetraplets;
        }
        resultCallRequests.push([
            key,
            {
                serviceId: callRequest.service_id,
                functionName: callRequest.function_name,
                arguments: arguments_,
                tetraplets: tetraplets,
            },
        ]);
    }
    return {
        retCode: result.ret_code,
        errorMessage: result.error_message,
        data: result.data,
        nextPeerPks: result.next_peer_pks,
        callRequests: resultCallRequests,
    };
}
exports.deserializeAvmResult = deserializeAvmResult;
/**
 * Utility function which serializes AVM args and passed them into AVM returning interpreter result.
 * Call to AVM is delegated to a function which must be provided by user.
 * It might be either synchronous or asynchronous (returning a promise)
 * @param fn - delegated call to AVM
 * @param initPeerId - peer ID which initialized particle
 * @param currentPeerId - peer ID which is currently executing the particle
 * @param air - particle's air script as string
 * @param prevData - particle's prev data as raw byte array
 * @param data - particle's data as raw byte array
 * @param callResults - array of tuples [callResultKey, callResult]
 * @returns structured InterpreterResult
 */
function callAvm(fn, runParams, air, prevData, data, callResults) {
    return __awaiter(this, void 0, void 0, function () {
        var avmArg, rawResult;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    avmArg = serializeAvmArgs(runParams, air, prevData, data, callResults);
                    return [4 /*yield*/, fn(avmArg)];
                case 1:
                    rawResult = _a.sent();
                    return [2 /*return*/, deserializeAvmResult(rawResult)];
            }
        });
    });
}
exports.callAvm = callAvm;
