// @fluencelabs/fluence/dist/__test__/_aqua/marine-js.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.call = exports.registerGreeting = void 0;
var v4_1 = require("../../internal/compilerSupport/v4");
function registerGreeting() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    (0, v4_1.registerService$$)(args, {
        defaultServiceId: 'greeting',
        functions: {
            tag: 'labeledProduct',
            fields: {
                greeting: {
                    tag: 'arrow',
                    domain: {
                        tag: 'labeledProduct',
                        fields: {
                            name: {
                                tag: 'scalar',
                                name: 'string',
                            },
                        },
                    },
                    codomain: {
                        tag: 'unlabeledProduct',
                        items: [
                            {
                                tag: 'scalar',
                                name: 'string',
                            },
                        ],
                    },
                },
                greeting_record: {
                    tag: 'arrow',
                    domain: {
                        tag: 'nil',
                    },
                    codomain: {
                        tag: 'unlabeledProduct',
                        items: [
                            {
                                tag: 'struct',
                                name: 'GreetingRecord',
                                fields: {
                                    num: {
                                        tag: 'scalar',
                                        name: 'i32',
                                    },
                                    str: {
                                        tag: 'scalar',
                                        name: 'string',
                                    },
                                },
                            },
                        ],
                    },
                },
            },
        },
    });
}
exports.registerGreeting = registerGreeting;
function call() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    var script = "\n                    (xor\n                     (seq\n                      (seq\n                       (seq\n                        (seq\n                         (seq\n                          (call %init_peer_id% (\"getDataSrv\" \"-relay-\") [] -relay-)\n                          (call %init_peer_id% (\"getDataSrv\" \"arg\") [] arg)\n                         )\n                         (call %init_peer_id% (\"greeting\" \"greeting\") [arg] res1)\n                        )\n                        (call %init_peer_id% (\"greeting\" \"greeting\") [res1] res2)\n                       )\n                       (call %init_peer_id% (\"greeting\" \"greeting\") [res2] res3)\n                      )\n                      (xor\n                       (call %init_peer_id% (\"callbackSrv\" \"response\") [res3])\n                       (call %init_peer_id% (\"errorHandlingSrv\" \"error\") [%last_error% 1])\n                      )\n                     )\n                     (call %init_peer_id% (\"errorHandlingSrv\" \"error\") [%last_error% 2])\n                    )\n    ";
    return (0, v4_1.callFunction$$)(args, {
        functionName: 'call',
        arrow: {
            tag: 'arrow',
            domain: {
                tag: 'labeledProduct',
                fields: {
                    arg: {
                        tag: 'scalar',
                        name: 'string',
                    },
                },
            },
            codomain: {
                tag: 'unlabeledProduct',
                items: [
                    {
                        tag: 'scalar',
                        name: 'string',
                    },
                ],
            },
        },
        names: {
            relay: '-relay-',
            getDataSrv: 'getDataSrv',
            callbackSrv: 'callbackSrv',
            responseSrv: 'callbackSrv',
            responseFnName: 'response',
            errorHandlingSrv: 'errorHandlingSrv',
            errorFnName: 'error',
        },
    }, script);
}
exports.call = call;
