// @fluencelabs/fluence/dist/internal/_aqua/node-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerNodeUtils = void 0;
var v4_1 = require("../../internal/compilerSupport/v4");
function registerNodeUtils() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    (0, v4_1.registerService$$)(args, {
        defaultServiceId: 'node_utils',
        functions: {
            tag: 'labeledProduct',
            fields: {
                read_file: {
                    tag: 'arrow',
                    domain: {
                        tag: 'labeledProduct',
                        fields: {
                            path: {
                                tag: 'scalar',
                                name: 'string',
                            },
                        },
                    },
                    codomain: {
                        tag: 'unlabeledProduct',
                        items: [
                            {
                                tag: 'struct',
                                name: 'ReadFileResult',
                                fields: {
                                    content: {
                                        tag: 'option',
                                        type: {
                                            tag: 'scalar',
                                            name: 'string',
                                        },
                                    },
                                    error: {
                                        tag: 'option',
                                        type: {
                                            tag: 'scalar',
                                            name: 'string',
                                        },
                                    },
                                    success: {
                                        tag: 'scalar',
                                        name: 'bool',
                                    },
                                },
                            },
                        ],
                    },
                },
            },
        },
    });
}
exports.registerNodeUtils = registerNodeUtils;
// Functions
