// @fluencelabs/fluence/dist/internal/_aqua/services.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerSig = void 0;
var v4_1 = require("../../internal/compilerSupport/v4");
function registerSig() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    (0, v4_1.registerService$$)(args, {
        defaultServiceId: 'sig',
        functions: {
            tag: 'labeledProduct',
            fields: {
                get_peer_id: {
                    tag: 'arrow',
                    domain: {
                        tag: 'nil',
                    },
                    codomain: {
                        tag: 'unlabeledProduct',
                        items: [
                            {
                                tag: 'scalar',
                                name: 'string',
                            },
                        ],
                    },
                },
                sign: {
                    tag: 'arrow',
                    domain: {
                        tag: 'labeledProduct',
                        fields: {
                            data: {
                                tag: 'array',
                                type: {
                                    tag: 'scalar',
                                    name: 'u8',
                                },
                            },
                        },
                    },
                    codomain: {
                        tag: 'unlabeledProduct',
                        items: [
                            {
                                tag: 'struct',
                                name: 'SignResult',
                                fields: {
                                    error: {
                                        tag: 'option',
                                        type: {
                                            tag: 'scalar',
                                            name: 'string',
                                        },
                                    },
                                    signature: {
                                        tag: 'option',
                                        type: {
                                            tag: 'array',
                                            type: {
                                                tag: 'scalar',
                                                name: 'u8',
                                            },
                                        },
                                    },
                                    success: {
                                        tag: 'scalar',
                                        name: 'bool',
                                    },
                                },
                            },
                        ],
                    },
                },
                verify: {
                    tag: 'arrow',
                    domain: {
                        tag: 'labeledProduct',
                        fields: {
                            signature: {
                                tag: 'array',
                                type: {
                                    tag: 'scalar',
                                    name: 'u8',
                                },
                            },
                            data: {
                                tag: 'array',
                                type: {
                                    tag: 'scalar',
                                    name: 'u8',
                                },
                            },
                        },
                    },
                    codomain: {
                        tag: 'unlabeledProduct',
                        items: [
                            {
                                tag: 'scalar',
                                name: 'bool',
                            },
                        ],
                    },
                },
            },
        },
    });
}
exports.registerSig = registerSig;
// Functions
